### DISCUSSION TITLE
Technical nits: Misc

### DISCUSSION THREAD

**Discussion by QuineDot** [google/comprehensive-rust#82] [Ideas]
[You make a direct comparison between Rust and C++ references,](https://google.github.io/comprehensive-rust/basic-syntax/references.html) but [Rust references are more similar to pointers in many ways.](https://stackoverflow.com/questions/64167637/is-the-concept-of-reference-different-in-c-and-rust)  I see people coming from C++ and being confused by this on the forums, so making the direct comparison may be counter-productive.

---

[The compile time guarantees](https://google.github.io/comprehensive-rust/why-rust/compile-time.html) are guaranteed for safe Rust with sound dependencies.  Worth pointing out since you have a section on `unsafe`.  Most of them are UB if they do get compiled.

There are no `NULL` *references*, but [there are `NULL` pointers.](https://doc.rust-lang.org/std/ptr/fn.null.html)

(More a suggestion:) Although foot-noted, the memory leak seemed out-of-place to me in this list.  Perhaps the `Mutex` lock too.  Both, I guess, fall under RAII.

---

[Here](https://google.github.io/comprehensive-rust/generics/trait-objects.html) you say
> you need a trait object if you want to return different values implementing a trait

But what you actually mean is "if you want to return different *types* implementing a trait".

The rest of this is suggestion territory, but [here](https://google.github.io/comprehensive-rust/generics/impl-trait.html) you should point out how different the argument position (APIT) and return position (RPIT) uses are.
* APIT is basically a generic parameter on the function, but the type becomes opaque
  * You don't need APIT to *accept* unnameable types in generics, e.g. to be passed a closure
  * The *caller* can choose *any type* that meets the bounds
* RPIT is opaque, but *still a single type* (for a given set of parameter *inputs*)
  * I.e. you can't return different types in an `if/else`
  * The *function writer* chooses the (*single*) type that gets returned

And then when you get to returning `Box<dyn Trait>` you can compare and contrast with RPIT.

---

[You say](https://google.github.io/comprehensive-rust/enums/sizes.html)
> Rust enums are packed tightly, 

But they're not really packed, it's more that the variants can share storage because it's a tagged union; only one variant is "active" at a time.   (And [`repr(packed)`](https://doc.rust-lang.org/nomicon/other-reprs.html#reprpacked) is something else entirely.)

Also, the layout (size optimization) is only guaranteed in very specific circumstances.  Perhaps mention in general that most Rust data structures have no layout guarantees.  [You mention niche optimization here too;](https://google.github.io/comprehensive-rust/std/box-niche.html) that *almost* meets [the guaranteed case](https://github.com/rust-lang/rust/pull/60300) (as far as I'm aware of it existing), but not quite.  So the niche optimization is not guaranteed for the example.

---

[You say](https://google.github.io/comprehensive-rust/exercises/day-1/iterators-and-ownership.html#iterator)
> The Iterator trait simply says that you can call next until you get None back

First, [there are 74 more other methods *so far*,](https://doc.rust-lang.org/std/iter/trait.Iterator.html) so it says a lot more than that.

But also, there's nothing that keeps you from calling `next` after you get a `None` back.  And unless your iterator is [fused,](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fuse) nothing that says you'll never get a `Some(_)` after you see a `None`.

---

[You say](https://google.github.io/comprehensive-rust/ownership/move-semantics.html)
> There is always exactly one variable binding which owns a value.

But [`Rc`](https://doc.rust-lang.org/std/rc/index.html) and [`Arc`](https://doc.rust-lang.org/std/sync/struct.Arc.html) are often referred to as "shared owneship".  Even more nit-picky, a `Vec` (say) can be in a single variable binding, but owns multiple values.

I don't have a good suggestion on how to word all this while getting across your original point and not being confusing though.

**Comment by mgeisler** [comment]
Hi @QuineDot, I converted this to a discussion from #69 — discussions have _threads_, so we can discuss each point in it's own little tree. It's like we've invented Usenet or mailing lists all over again :smile:

**Comment by mgeisler** [comment]
> [You say](https://google.github.io/comprehensive-rust/ownership/move-semantics.html)
> 
> > There is always exactly one variable binding which owns a value.
> 
> But [`Rc`](https://doc.rust-lang.org/std/rc/index.html) and [`Arc`](https://doc.rust-lang.org/std/sync/struct.Arc.html) are often referred to as "shared owneship". Even more nit-picky, a `Vec` (say) can be in a single variable binding, but owns multiple values.
> 
> I don't have a good suggestion on how to word all this while getting across your original point and not being confusing though.

Interesting point about the `Vec`... I guess my mental model is something like this:

1. There is a single owner of each value
2. A value can become part of a compound type: a struct field, an array element, or put on the heap like in the case of a `Vec`

With that, I guess a `Option<T>` doesn't _own_ the `T` value. Instead, the `T` values is _part of_ the `Option`. Likewise with a `Vec<T>` where the `T`s are _part of` the `Vec`. I don't know if that makes sense?

I like this way of thinking since it circumvents the whole question about what it means for a type (`Vec`) to own another type (`T`).

I also don't think we should try to write this out in full details — it's something the instructor will have to discuss with the students during class.

**Comment by mgeisler** [comment]
> [You say](https://google.github.io/comprehensive-rust/exercises/day-1/iterators-and-ownership.html#iterator)
> 
> > The Iterator trait simply says that you can call next until you get None back
> 
> First, [there are 74 more other methods _so far_,](https://doc.rust-lang.org/std/iter/trait.Iterator.html) so it says a lot more than that.
> 
> But also, there's nothing that keeps you from calling `next` after you get a `None` back. And unless your iterator is [fused,](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fuse) nothing that says you'll never get a `Some(_)` after you see a `None`.

That's a fair point... I should not have used the word _simply_ here :smile: We do talk about how you implement one method (`next`) and get a ton of functionality for free. We can probably rephrase this a bit to make it clearer and more helpful.
